components/suggest/src/schema.rs

```diff
@@ -149,15 +158,217 @@ impl ConnectionInitializer for SuggestConnectionInitializer {
     }
 
-    fn upgrade_from(&self, _db: &Transaction<'_>, version: u32) -> open_database::Result<()> {
+    fn upgrade_from(&self, tx: &Transaction<'_>, version: u32) -> open_database::Result<()> {
         match version {
             1..=15 => {
                 // Treat databases with these older schema versions as corrupt,
                 // so that they'll be replaced by a fresh, empty database with
                 // the current schema.
                 Err(open_database::Error::Corrupt)
             }
+            16 => {
+                tx.execute(
+                    "
+                    CREATE TABLE dismissed_suggestions (
+                        url_hash INTEGER PRIMARY KEY
+                    ) WITHOUT ROWID;",
+                    (),
+                )?;
+                Ok(())
+            }
             _ => Err(open_database::Error::IncompatibleVersion(version)),
         }
     }
 }
+
+/// Clears the database, removing all suggestions, icons, and metadata.
+pub fn clear_database(db: &Connection) -> rusqlite::Result<()> {
+    db.execute_batch(
+        "
+        DELETE FROM meta;
+        DELETE FROM suggestions;
+        DELETE FROM icons;
+        DELETE FROM yelp_subjects;
+        DELETE FROM yelp_modifiers;
+        DELETE FROM yelp_location_signs;
+        DELETE FROM yelp_custom_details;
+        ",
+    )
+}
+
+fn define_functions(c: &Connection) -> rusqlite::Result<()> {
+    c.create_scalar_function(
+        "hash",
+        1,
+        FunctionFlags::SQLITE_UTF8 | FunctionFlags::SQLITE_DETERMINISTIC,
+        sql_fns::hash,
+    )?;
+    Ok(())
+}
+
+mod sql_fns {
+    use md5::{Digest, Md5};
+    use rusqlite::{functions::Context, types::ValueRef, Error, Result};
+
+    #[inline(never)]
+    pub fn hash(ctx: &Context<'_>) -> Result<Option<i64>> {
+        match ctx.len() {
+            1 => {
+                let raw = ctx.get_raw(0);
+                // This is a deterministic function, which means sqlite
+                // does certain optimizations which means hash() may be called
+                // with a null value even though the query prevents the null
+                // value from actually being used. As a special case, we return
+                // null when the input is NULL. We return NULL instead of zero
+                // because the hash columns are NOT NULL, so attempting to
+                // actually use the null should fail.
+                if raw == ValueRef::Null {
+                    return Ok(None);
+                }
+                match raw.as_str() {
+                    Ok(s) => {
+                        // Hash using Md5, then return the first 64 bits since SQLite uses 64-bit
+                        // integers.
+                        let mut hasher = Md5::new();
+                        hasher.update(s.as_bytes());
+                        let hash_value = i128::from_le_bytes(hasher.finalize().into());
+                        Ok(Some(hash_value as i64))
```

Comment:
Nit: let's grab the leading 64 bits explicitly. Using `as` to cast from a larger integer to a smaller integer will truncate which could yield weird negative results.

```rs
    println!("{}", (i128::MAX >> 64) as i64);  // => 9223372036854775807

    println!("{}", i128::MAX as i64);               // => -1
    println!("{}", (i128::MAX - 1) as i64);       // => -2
    println!("{}", (i128::MAX - 2) as i64);      // => -3
    println!("{}", (i128::MAX - 3) as i64);      // => -4
```

 ```suggestion
                         Ok(Some((hash_value >> 64) as i64))
```

Reply:
I updated the code to do this, but I think there still could be negatives (for example `u128::MAX` gets converted to `-1`).  We could shift by 65 bits to avoid this, but I'd rather keep the extra hash bit and live with potentially negative numbers.  WDYT?